modules/@apostrophecms/rich-text-widget/index.js

```diff
@@ -567,10 +585,87 @@ module.exports = {
             const right = content.indexOf('>', left);
             const nextLeft = content.indexOf('<', right);
             if ((right !== -1) && (nextLeft !== -1)) {
-              content = content.substr(0, right + 1) + self.apos.util.escapeHtml(doc.title) + content.substr(nextLeft);
+              content = content.substring(0, right + 1) + self.apos.util.escapeHtml(doc.title) + content.substring(nextLeft);
+            }
+          }
+        }
+        return content;
+      },
+      // Quickly replaces inline image placeholder URLs with
+      // actual, SEO-friendly URLs based on `widget._relatedDocs`
+      linkImages(widget, content) {
+        if (widget._edit) {
+          return content;
+        }
+        // "Why no regexps?" We need to do this as quickly as we can.
+        // indexOf and lastIndexOf are much faster.
+        let i;
+        for (const doc of (widget._relatedDocs || [])) {
+          let offset = 0;
+          while (true) {
```

Comment:
Is `while` used for performance here?

Reply:
Yes. This is brought forward from carefully tuned code in the A2 permalinks module. This is extra work done at read time, and Apostrophe is read-optimized, so getting a little bit micro with the optimizations is warranted, even at some cost in readability.